feat(tracing): Add hook for trace sampling function to SDK options

lobsterkatie opened the pull request:
Currently, tracing is either on, with a static sample rate for all transactions, or off. This doesn't let users either

  a) sample different transactions at different rates or
  b) filter out certain transactions all together.

This PR introduces a new SDK option, `tracesSampler`, whose value is a callback with the signature

`tracesSampler(samplingContext: SamplingContext): number | boolean`,

where `samplingContext` is a combination of automatically-provided data (varying by SDK/integration) and data passed by the user to `startTransaction()` as an optional third parameter. Given that data, the `tracesSampler` function can then compute the appropriate sample rate and return it to the sampling function.

Example use cases for variable sample rate:

- Collect traces for all enterprise customers but only a fraction of regular customers
- Collect a higher percentage of traces on rarely-visited pages, to ensure a decent sample size
- Collect more traces on a page you recently revamped compared to one whose code you already have well-profiled

Example use cases for filtering:

- Don't trace pre-flight CORS requests
- Don't trace health check requests
- Don't collect traces from users with a certain browser extension enabled, because it's known to break things
- Don't trace navigation to deprecated parts of your web app

(A user can, of course, combine these ideas as well, returning a variable sample rate for some traces and 0 to filter out others.)

In order to support this addition, a few other changes needed to be made:

- Move method for extracting node request data to `@sentry/utils`
- Break up `misc.ts` to eliminate circular dependencies within `@sentry/utils`
- Tweak some types related to domains and request data, and polyfill `WorkerLocation` type


Still to do:

- [x] Write a better PR description
- [x] Decide what can (and should) go into the `SampleContext` object, and put it there
- [X] Try it out
- [X] Write tests
- [x] Deal with circular dependency between packages
- [ ] Fix ~linting and~ size check errors
- [x] Use `isNaN` in rate validator
- [x] Make sampler take precedence over parent's decision, pass parent's decision in sampleContext
- [ ] Update docs (including mentioning what data is automatically passed in `sampleContext` depending on the SDK)

kamilogorek commented on packages/gatsby/gatsby-browser.js:
```diff
@@ -34,7 +34,8 @@ exports.onClientEntry = function(_, pluginParams) {
       dsn: __SENTRY_DSN__,
       ...pluginParams,
-      tracesSampleRate,
+      tracesSampleRate: pluginParams.tracesSampleRate,
```
These 2 lines are redundant, as `...pluginParams` will include them implicitly

lobsterkatie replied:
Shoot, you're right - we're no longer modifying the value, so it doesn't need to be separate like it used to be. Fixed.

kamilogorek commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -19,25 +20,68 @@ function traceHeaders(this: Hub): { [key: string]: string } {
 
 /**
- * Use RNG to generate sampling decision, which all child spans inherit.
+ * Use sample rate (given in options as either a constant or ) along with a random number generator to make a sampling
```
> constant or

? 😅

lobsterkatie replied:
🤔 

Good question.

lobsterkatie replied:
Oh, heh. Incomplete editing. (Moved that parenthetical to its own sentence on the next line, but missed getting rid of it once I had.)

Thanks for catching. Fixed.

kamilogorek commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -19,25 +20,68 @@ function traceHeaders(this: Hub): { [key: string]: string } {
  */
 function sample<T extends Transaction>(hub: Hub, transaction: T): T {
   const client = hub.getClient();
+  const options = (client && client.getOptions()) || {};
```
This could be written as:

```js
const client = hub.getClient();

// I assume if there's no client, there's no point in doing the rest of operations, as they all depend on the existence of options anyway
if (!client) {
  return transaction;
}

// `getOptions` always return at least an empty object. It's impossible to initialize the client otherwise.
const options = client.getOptions();

lobsterkatie replied:
True. It replaces 1 line with 5 lines, though...
![image](https://user-images.githubusercontent.com/14812505/90264014-4e6a1800-de05-11ea-9a3d-180d6cd7f0af.png)

There is something nice about acknowledging that not having a client matters, though. What about this?
![image](https://user-images.githubusercontent.com/14812505/90264282-b4569f80-de05-11ea-9a94-64c46457fef4.png)

kamilogorek commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -19,25 +20,68 @@ function traceHeaders(this: Hub): { [key: string]: string } {
-    // if true = we want to have the transaction
-    // if false = we don't want to have it
-    // Math.random (inclusive of 0, but not 1)
+    let sampleRate;
```
```js
const sampleContext = {}; // todo
const sampleRate = typeof options.tracesSampler === 'function' ? options.tracesSampler(sampleContext) : options.tracesSampleRate;
```

lobsterkatie replied:
Nice. Done.

kamilogorek commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -19,25 +20,68 @@ function traceHeaders(this: Hub): { [key: string]: string } {
+
+    // if the function returned either 0 or null, it's a sign the transaction should be dropped
+    if (!sampleRate) {
+      logger.log('Discarding trace because tracesSampler returned 0 or null');
```
```suggestion
      logger.log('Discarding trace because tracesSampler returned 0 or null or sampleRate was set to 0');
```

lobsterkatie replied:
✅

kamilogorek commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -19,25 +20,68 @@ function traceHeaders(this: Hub): { [key: string]: string } {
 
+  // at this point we know we're keeping the transaction, whether because of an inherited decision or because it got
+  // lucky with the dice roll
+  const experimentsOptions = (client && client.getOptions()._experiments) || {};
```
If my patch above will be applied, then this can be changed to `options._experiments || {}`

lobsterkatie replied:
You're right. Actually, even without your patch the check was unnecessary. Fixed.

kamilogorek commented on packages/types/src/options.ts:
```diff
@@ -144,4 +145,13 @@ export interface Options {
+   *
+   * @returns A sample rate or `null` to drop the trace.
+   */
+  tracesSampler?(traceContext: SampleContext): number | null;
```
Tbh I'm not sure why we accept `null` here. It makes things unnecessarily more complicated.
`0` is exactly the same as `null` from technical standpoint here.

lobsterkatie replied:
It's true - it's not totally necessary. I was going for the parallel with `beforeSend`, but if you think that's not worth it, I can take it out.

lobsterkatie replied:
Update: Conversation elsewhere has convinced me you're right. Will take it out.

AbhiPrasad commented on packages/gatsby/gatsby-browser.js:
```diff
@@ -1,6 +1,7 @@
 exports.onClientEntry = function(_, pluginParams) {
   require.ensure(['@sentry/react'], function(require) {
     const Sentry = require('@sentry/react');
+    const hasTracingEnabled = require('@sentry/utils').hasTracingEnabled;
```
Can we make sure `@sentry/utils` is in `require.ensure`, and explicitly put it in the deps if we are going to use it?

lobsterkatie replied:
Sure. TIL about `require.ensure()`. LMK if there's more I need to do here, besides adding it to the array and to `package.json`.

rhcarvalho commented on packages/gatsby/gatsby-browser.js:
```diff
@@ -1,15 +1,15 @@
   const integrations = [...(pluginParams.integrations || [])];
 
-  if (tracesSampleRate && !integrations.some(ele => ele.name === 'BrowserTracing')) {
+  if (Utils.hasTracingEnabled(pluginParams) && !integrations.some(ele => ele.name === 'BrowserTracing')) {
```
Perhaps `hasTracingEnabled` would be best suited to live in `@sentry/tracing`. As we can see here, `@sentry/gatsby` did not depend on `@sentry/utils` and now it does just because of... tracing. We shouldn't need to pull in all of utils for a single tracing-related helper.

lobsterkatie replied:
Fair point. Done.

kamilogorek commented on packages/utils/src/misc.ts:
```diff
@@ -510,3 +375,12 @@ export function addContextToFrame(lines: string[], frame: StackFrame, linesOfCon
+ * Tracing is enabled when at least one of `tracesSampleRate` and `tracesSampler` is defined in the SDK config.
+ */
+export function hasTracingEnabled(options: Options): boolean {
+  return !!options.tracesSampleRate || !!options.tracesSampler;
```
```suggestion
  return 'tracesSampleRate' in options || 'tracesSampler' in options;
```

It's more descriptive imo :)

lobsterkatie replied:
You're right. :-)

Fixed.

rhcarvalho reviewed (changes requested):
My main concern here is whether we can have a clear and consistent type def for SamplingContext, not an arbitrary mapping of strings to anything.
That should make documentation easier and usage easier.

Looking at the impl. I'm also not convinced it has to be added to `startTransaction` as a new argument. If we consider Hub is an interface, changing method signatures is a breaking change.
Perhaps this is just because there's no usage example yet -- tests would help clarify what are the use cases being addressed.

lobsterkatie commented on packages/hub/src/hub.ts:
```diff
@@ -456,22 +457,25 @@ export function getCurrentHub(): Hub {
+  const sentry = getMainCarrier().__SENTRY__;
+
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+  return sentry && sentry.extensions && sentry.extensions.domain && (sentry.extensions.domain as any).active;
```
Fixed the types so no linter rules are needed.

lobsterkatie commented on packages/tracing/test/hub.test.ts:
```diff
@@ -1,56 +1,248 @@
+
+        expect(tracesSampler).toHaveBeenCalledWith(expect.objectContaining({ location: dogParkLocation }));
+      });
+    }); // end describe('defaultSampleContext')
```
TIL `cmd+shift+\`.

lobsterkatie commented on packages/tracing/test/hub.test.ts:
```diff
@@ -1,56 +1,248 @@
 addExtensionMethods();
 
 describe('Hub', () => {
+  beforeEach(() => {
+    jest.spyOn(logger, 'warn');
+    jest.spyOn(logger, 'log');
+    jest.spyOn(utilsModule, 'isNodeEnv');
+
+    // NB: Upon refactoring, this spy was no longer needed. Leaving it in as an excuse to leave in the note above, so
+    // that it can save future folks the headache.
+    jest.spyOn(hubModule, 'getActiveDomain');
+  });
```
Done.

lobsterkatie commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -20,34 +29,150 @@ function traceHeaders(this: Hub): { [key: string]: string } {
+    const globalObject = getGlobalObject();
+
+    if ('location' in globalObject) {
+      // we take a copy of the location object rather than just a reference to it in case there's a navigation or
+      // redirect in the instant between when the transaction starts and when the sampler is called
```
Good question. I'd say that the chances are slim that there's a race condition anyway, and that this roughly halves the chances of it happening. That said, that's not the same as fixing it, and the copying could just as easily be justified by not wanting to give folks an opportunity to muck with the real `location` object and accidentally _cause_ a navigation, so I can switch to that justification (or no justification) - whatever you all think is best.

@kamilogorek, this was your suggestion. Do you have thoughts here?

lobsterkatie commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -20,34 +29,150 @@ function traceHeaders(this: Hub): { [key: string]: string } {
+    const domain = getActiveDomain();
+
+    if (domain) {
+      // for all node servers that we currently support, we store the incoming request object (which is an instance of
+      // http.IncomingMessage) on the domain
+
+      // the domain members are stored as an array, so our only way to find the request is to iterate through the array
+      // and compare types
```
It is in fact the existing behavior, and my understanding is that scope/thread/concurrency safety the reason we still use the deprecated `domain` package, because we don't have another way to accomplish that. @kamilogorek is the expert here, though.

lobsterkatie commented on packages/types/src/options.ts:
```diff
@@ -84,7 +85,7 @@ export interface Options {
    * 1.0 = 100% chance of a given trace being sent (send all traces)
    *
-   * Default: 0.0
+   * Can be omitted without disabling tracing if `tracesSampler` is defined. If neither is defined, tracing is disabled.
```
The linter insists on having all properties before any methods, so I moved `tracesSampleRate` to be the last property and `tracesSampler` to be the first method, so they're together. I also reworded the description. LMK if you feel like it's clearer now.

lobsterkatie commented on packages/tracing/src/hubextensions.ts:
```diff
@@ -20,34 +29,150 @@ function traceHeaders(this: Hub): { [key: string]: string } {
+      const request = domain.members.find((member: any) => isInstanceOf(member, requestType));
+
+      if (request) {
+        defaultSampleContext.request = extractNodeRequestData(request);
```
The idea is to give people as much data as is easily grabbed. Naturally, that data is going to be different depending on the context - browser window, service worker, or node server.

lobsterkatie commented:
@rhcarvalho I've tried to address your concerns re: specificity (or lack thereof) in the definition of the `SampleContext` type.

1) The type now consists of two halves, `CustomSampleContext` and added attributes (the defaults) so it's clear that we're combining two data sources. (I originally had the defaults in their own type, `DefaultSampleContext`, but I moved them to the main `SampleContext` type so that when you examine the `SampleContext` type (in VSCode or other), the default stuff shows up right there.)

2) The defaults now have named keys so it's clear what gets passed automatically.

3) The third parameter in `startTransaction` is now called `customSampleContext`, to make it clear that it's coming from the user.

LMK what you think.

```typescript
/**
 * Context data passed by the user when starting a transaction, to be used by the tracesSampler method.
 */
export interface CustomSampleContext {
  [key: string]: any;
}

/**
 * Data passed to the `tracesSampler` function, which forms the basis for whatever decisions it might make.
 *
 * Adds default data to data provided by the user. See {@link Hub.startTransaction}
 */
export interface SampleContext extends CustomSampleContext {
  /**
   * Object representing the URL of the current page or worker script. Passed by default in a browser or service worker
   * context.
   */
  location?: Location | WorkerLocation;

  /**
   * Object representing the incoming request to a node server. Passed by default when using the TracingHandler.
   */
  request?: ExtractedNodeRequestData;
}
```

lobsterkatie commented on packages/types/src/options.ts:
```diff
@@ -144,4 +145,17 @@ export interface Options {
+   *
+   * Can be defined in place of `tracesSampleRate`. If neither is defined, tracing is disabled.
+   *
+   * `startTransaction` or `startIdleTransaction` accept a `sampleContext` argument, which is then passed to this
```
> It is a bit of a maintenance burden when we add this sort of cross documentation. I understand it makes sense now, but it is hard to keep consistent.

Switched to linking to the `startTransaction` docstring.

> Better to document SampleContext in the jsdoc for SampleContext.
>
> People looking at the API docs can look for where SampleContext appears -- the results would include startTransaction and any other place it may be used.

See https://github.com/getsentry/sentry-javascript/pull/2820#issuecomment-685826016.

> And we've been considering startIdleTransaction an internal thing for now, so it should not be prominent here.

Removed mention of it, as well as the ability to pass custom context to it at all. As you say, it's internal, so user-provided context doesn't make sense.

lobsterkatie commented on packages/types/src/transaction.ts:
```diff
@@ -48,3 +48,11 @@ export interface Transaction extends TransactionContext, Span {
+ * Combination of default values (which differ per SDK/integration) and data passed to `startTransaction`.
+ */
+export interface SampleContext {
+  [key: string]: any;
```
See https://github.com/getsentry/sentry-javascript/pull/2820#issuecomment-685826016.

lobsterkatie commented on packages/gatsby/README.md:
```diff
@@ -50,7 +50,10 @@ To enable Tracing support, supply the `tracesSampleRate` to the options and make
-          tracesSampleRate: 1, // this is just to test, you should lower this in production
+
+          // A rate of 1 means all traces will be sent, so it's good for testing.
+          // In production, you'll likely want to either choose a lower rate or use `tracesSampler` instead.
```
Have updated the gatsby readme and will come back to the question of linking when I write the docs for this.

HazAT commented on packages/hub/src/interfaces.ts:
```diff
@@ -20,9 +21,23 @@ export interface Carrier {
      */
-    // eslint-disable-next-line @typescript-eslint/ban-types
-    extensions?: { [key: string]: Function };
+    extensions?: {
+      /** Hack to prevent bundlers from breaking our usage of the domain package in the cross-platform Hub package */
+      domain?: typeof domain & {
+        /**
+         * The currently active domain. This is part of the domain package, but for some reason not declared in the
+         * package's typedef.
+         */
+        active?: domain.Domain;
+      };
+    } & {
+      /** Extension methods for the hub, which are bound to the current Hub instance */
+      // eslint-disable-next-line @typescript-eslint/ban-types
+      [key: string]: Function;
+    };
```
H: Are we sure that doesn't break in Browser environments?

lobsterkatie replied:
I will check again, for sure, but the change here isn't a behavior change but merely a TS one. We've always put the domain there, it's just that before we didn't include it in the type so we had to do a lot of `eslint-disable @no-unsafe-member-access` and `eslint-disable @no-explicit-any`. My understanding is that TS stuff all gets stripped?

Anyway, will check and update here.

lobsterkatie replied:
UPDATE: Seems to build just fine, with no special accommodation in my set up.

MhMadHamster replied:
@lobsterkatie It seems like these changes broke `@sentry/browser` for TS users by requiring node types 😕 issue: https://github.com/getsentry/sentry-javascript/issues/2927

aleclarson replied:
See #3050 for a fix.

HazAT reviewed (approved):
Otherwise LGTM, good PR with a lot of refactorings 👍

HazAT reviewed (commented):
One more thing, why did we decide not to provide the `Transaction` in the `SamplingContext`?

lobsterkatie commented:
> One more thing, why did we decide not to provide the Transaction in the SamplingContext?

Okay, switched from passing `parentSampled` to passing `transactionContext`. I didn't pass the transaction itself because I felt like until it is returned from `startTransaction()`, it shouldn't be able to be mucked with from the outside (like by `tracesSampler()`). 

That said, I think inheritance is a common and important enough concern that I'd like to be able to do better than "this existing decision _probably_ came from the parent, but honestly, we don't know for sure." But in the interests of getting this through, I'll split my proposed solution there into a separate PR.

aleclarson commented on packages/hub/src/interfaces.ts:
```diff
@@ -20,9 +21,23 @@ export interface Carrier {
+         * The currently active domain. This is part of the domain package, but for some reason not declared in the
+         * package's typedef.
+         */
+        active?: domain.Domain;
```
We shouldn't be using `@types/node` in an isomorphic package.

dvins commented:
@kamilogorek @lobsterkatie This feature introduced #2971 some months ago, the only solution for many is to down level the package version. It appears to have fallen off the radar despite a number of us customers having this problem. We're now pushing on three months with no feedback, let alone a fix.

dvins commented:
Can someone front Sentry at least provide some feedback, even if it's as trite as 'We have no plans to resolve this so customers like us can make some decisions? @kamilogorek @lobsterkatie

rhcarvalho commented:
@dvins please try https://github.com/getsentry/sentry-javascript/releases/tag/6.0.4. #2971 is now closed, if there's anything else preventing you from upgrading please let us know in a new issue. Thanks!
